draft-normington-jsonpath-latest.xml

```diff
@@ -104,7 +107,75 @@
       </section>
     </section>
    <section numbered="true" toc="default">
-      <name>JSONPath Syntax and Semantics</name>         
+      <name>JSONPath Syntax and Semantics</name>
+         <section numbered="true" toc="default">
+            <name>Overview</name>
+            <t>A JSONPath is a string which matches zero or more nodes of any JSON document. A valid JSONPath conforms to the ABNF syntax defined by this document.
+            </t>
+         </section>
+         <section numbered="true" toc="default">
+            <name>Implementation</name>
+            <t>An implementation of this specification, from now on referred to simply as "an implementation", SHOULD takes two inputs, a JSONPath and a JSON document, and produce
+               a possibly empty list of nodes of the JSON document which match the JSONPath or an error (but not both).
+            </t>
+            <t>If there is no matching node and no error has occurred, an implementation MUST return an empty list of nodes.
+            </t>
+            <t>Syntax errors in the JSONPath SHOULD be detected before matching is attempted since they do not depend on the JSON document.
+               Therefore, an implementation SHOULD take a JSONPath and produce an optional syntax error and then,
+               if and only if an error was not produced, SHOULD take a JSON document and produce a list of matches or an error (but not both).
+            </t>
+            <t>Alternatively, an implementation MAY take a JSONPath and a JSON document and produce a list of matches or an optional error (but not both).
+            </t>
+            <t>For any implementation, if a syntactically invalid JSONPath is provided, the implementation MUST return an error.
+            </t>
+            <t>If a syntactially invalid JSON document is provided, an implementation MAY return an error or MAY return an empty list of matches.
+            </t>
+         </section>
+         <section numbered="true" toc="default">
+            <name>Syntax</name>
+            <t>Syntactically, a JSONPath consists of <tt>root</tt>, which matches the root node of a JSON document, followed by a possibly empty sequence of <em>matchers</em>.
+            </t>
+            <sourcecode type="abnf">
+jsonpath = root *matcher
+root = %x24              ; $ matches the root node of a document
+            </sourcecode>
+            <t>The syntax and semantics of each matcher is defined below.
+            </t>
+         </section>
+         <section numbered="true" toc="default">
+            <name>Semantics</name>
+            <t><tt>root</tt> not only matches the root node of the input document, but it produces as output a list consisting of one node: the input document.
```

Comment:
> matches the root node

If `$` is a matcher, then it can appear at any position in chain, `$[1]$`, for example. Maybe it's better to define `$` and `@` as _selectors_ and a JsonPath query as selector, followed by matchers sequence (possibly empty)?

Reply:
Yes, I was trying to avoid describing `$` as a matcher for the reason you point out. I guess saying that `$` _matches_ the root node gives the impression that `$` is a matcher (regardless of what the ABNF might say). Can you think of a better wording? If not, I'll try again...

Reply:
> Maybe it's better to define `$` and `@` as _selectors_ and a JsonPath query as selector, followed by matchers sequence (possibly empty)?

I was trying to stick pretty closely to the Proposal A grammar. Also, we need to prevent `@` at the start of a JSONPatth. But your point is well made and will come in useful once we get to filters. Do we really need to do more right now?

Reply:
> we need to prevent @ at the start of a JSONPath

In fact, we do not. Semantic of `@` is "current node", and current node at the start of JsonPath is obviously a root node, so there's no practical reason to avoid `@` at start.

But it's, I'd say, a brave movement and should be discussed more thoroughly. For now, I believe, it's enough to choose between "sequence of matchers" and "selector followed by optional sequence of matchers".

Reply:
I'd like us to adopt an approach for the first version of the internet draft of avoiding new constructs whenever possible as that will increase the migration cost for existing implementations and users. So I think in this PR we should stick with `$` followed by a possibly empty sequence of matchers. I'll reword the above text to avoid calling `$` a matcher and use the "selector" terminology instead, which we can then piggy-back upon nicely when we come to filters. Let me draft that and push the change for your perusal.

Reply:
> avoiding new constructs whenever possible

That's not a problem - we can define selectors with the following grammar:

```
selector       = root_selector / other_selector
root_selector  = "$"
other_selector = "@"
```

And then define queries like this, avoiding:

```
root_path   = root_selector [ matcher_list ]
filter_path = selector [ matcher_list ]
```

It's not the concrete grammar proposal, just showing an idea how to forbid `@` at the start of outer path.

Reply:
Thanks @remorhaz. I've used just what we need of this for the moment. I adopted the convention from the ABNF RFC of spelling compound grammar elements using "-" rather than "_".